pallets/ethereum-transaction/src/lib.rs

```diff
@@ -0,0 +1,138 @@
+// Copyright 2023 Centrifuge Foundation (centrifuge.io).
+//
+// This file is part of the Centrifuge chain project.
+// Centrifuge is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version (see http://www.gnu.org/licenses).
+// Centrifuge is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use cfg_traits::ethereum::EthereumTransactor;
+use ethereum::{LegacyTransaction, TransactionAction, TransactionSignature, TransactionV2};
+use fp_evm::CallOrCreateInfo;
+use frame_support::{
+	dispatch::{DispatchErrorWithPostInfo, PostDispatchInfo},
+	pallet_prelude::*,
+};
+use pallet_evm::GasWeightMapping;
+use sp_core::{H160, H256, U256};
+
+mod weights;
+
+pub use pallet::*;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub (super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config + pallet_ethereum::Config {
+		type EthereumOrigin: EnsureOrigin<Self::RuntimeOrigin, Success = H160>;
+	}
+
+	/// Storage for nonce.
+	#[pallet::storage]
+	pub(crate) type Nonce<T: Config> = StorageValue<_, U256, ValueQuery>;
+
+	impl<T: Config> EthereumTransactor for Pallet<T> {
+		fn call(
+			from: H160,
+			to: H160,
+			data: &[u8],
+			value: U256,
+			gas_price: U256,
+			gas_limit: U256,
+		) -> DispatchResultWithPostInfo {
```

Comment:
Do we want to propagate this and then charge for the fees in one of the callers of this OR do we want to charge all the necessary fees here, in this pallet?

Reply:
Upon further testing I noticed that some fees are actually charged by the [EVM runner](https://github.com/paritytech/frontier/blob/polkadot-v0.9.38/frame/evm/src/runner/stack.rs#L187).

Reply:
In this case, the above mentioned fees would be charged from the derived account, meaning that we'll have to ensure that the original caller will have sufficient funds to execute the EVM call.

Reply:
Thanks for the research. What would happen if the evm fails? Would it roll-back the complete transaction including possible EVM changes?

If this is the case, I would just assume we fail in runtime then.

Reply:
I will have to check that.

Reply:
> If there is an error before [this point](https://github.com/paritytech/frontier/blob/5644279da588baab7caa9f3a4eb69363e1606461/frame/evm/src/runner/stack.rs#L187) we will get back a `RunnerError` that contains a weight that we should charge. Otherwise, the total fee for the EVM transaction will be charged by the runner and we would only have to charge for reading the nonce.

Given this, would it make sense to charge for reading the nonce in this pallet? This would the allow us to just use a `DispatchResult` here and keep other parts free of fee-related logic. WDYT @mustermeiszer